docs/versioning.md

```diff
@@ -0,0 +1,567 @@
+"Master Version". The subsection that immediately follows describes the metadata
+key and version ID format in further detail.
+
+### Version ID and Metadata Key Format
+
+The version ID is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3 before sending a response to a request. S3 also
+decodes the hexadecimal string received from a request before sending to
+metadata to retrieve a particular version.
+
+The format of a `version_id` is:
+`ts` `rep_group_id` `seq_id`
+where:
+
+- `ts`: is the combination of epoch and an increasing number
+- `rep_group_id`: is the name of deployment(s) considered one unit used for
+   replication
+- `seq_id`: is supposed to be a unique value based on metadata information.
+
+The format of a key in metadata for a version is:
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the format described above as metadata orders keys
+  alphabetically
+
+An example of a key in metadata: `foo\01234567890000777PARIS 1234.123456`
+indicating that this specific version of `foo` was the `000777`th entry created
+during the epoch `1234567890` in the replication group `PARIS` with
+`1234.123456` as `seq_id`.
+
+### Master Version
+
+We store a copy of the latest version of an object's metadata using
+`object_name` as the key; this version is called the master version. The master
+versions of each object facilitates the standard GET operation, which would
+otherwise need to scan among the list of versions of an object for its latest
+version.
+
+The following table shows the layout of all versions of `foo` in the first
+example stored in the metadata (with dot `.` representing the null separator):
+
+| key | value |
+|-----|:-----:|
+| foo |   B   |
+| foo.v2 |   B   |
+| foo.v1 |   A   |
+
+### Metadata Versioning Options
+
+S3 Server sends instructions to the metadata engine about whether to create a
+new version or overwrite, retrieve, or delete a specific version by sending
+values for special options in PUT, GET, or DELETE calls to metadata. The
+metadata engine can also list versions in the database, which is used by S3
+to list object versions.
+
+These only describe the basic CRUD operations that the metadata engine can
+handle. How these options are used by the S3 API to generate and update versions
+involves more steps and is described more comprehensively in "Implementation of
+Bucket Versioning in API" and its subsections.
+
+Note: all operations (PUT and DELETE) that generate a new version of an object
+will return the `version_id` of the new version to the API.
+
+#### PUT
+
+- no options: original PUT operation, will update the master version
```

Comment:
I think it's a helpful comma :)
It means: "This is the original PUT operation. Will update the master version"